bindings/java/c/host.c

```diff
@@ -0,0 +1,847 @@
+        // using space reserved for negative numbers(see UnsignedLong for convenience).
+        // if this number is truly a signed 64bit number, then Java can't handle it
+        jnumber = (jlong)number;
+
+        // call java method
+        jobject jresult =
+            (*jenv)->CallStaticObjectMethod(jenv, host_class, method, jcontext_index, jnumber);
+
+        if (!jresult)
+        {
+            printf("JNI Error: invalid jresult in method %s\n", java_method_name);
+            detach(jenv);
+        }
+
+        evmc_bytes32* result_ptr =
+            (struct evmc_bytes32*)(*jenv)->GetDirectBufferAddress(jenv, jresult);
+        assert(result_ptr != NULL);
+        result = *result_ptr;
+    }
+    return result;
+}
+
+static void emit_log_fn(struct evmc_host_context* context,
+                        const evmc_address* address,
+                        const uint8_t* data,
+                        size_t data_size,
+                        const evmc_bytes32 topics[],
+                        size_t topics_count)
+{
+    const char java_method_name[] = "emit_log";
+    const char java_method_signature[] = "(I[B[BI[[BI)V";
+    JNIEnv* jenv = attach();
+    if (jenv != NULL)
+    {
+        jclass host_class;
+        jmethodID method;
+        jint jcontext_index;
+        jbyteArray jaddress;
+        jbyteArray jdata;
+        jobjectArray jtopics;
+
+        // get java class
+        host_class = (*jenv)->FindClass(jenv, "org/ethereum/evmc/Host");
+        if (!host_class)
+        {
+            printf("JNI Error: FindClass was unable to load the Host class\n");
+            detach(jenv);
+        }
+
+        // get java method
+        method =
+            (*jenv)->GetStaticMethodID(jenv, host_class, java_method_name, java_method_signature);
+        if (!method)
+        {
+            printf("JNI Error: GetStaticMethodID was unable to find method %s\n", java_method_name);
+            detach(jenv);
+        }
+
+        // set java method params
+
+        if (!context)
+        {
+            printf("JNI Error: context is NULL in method %s\n", java_method_name);
+            detach(jenv);
+        }
+        jcontext_index = context->index;
+
+        jaddress = (*jenv)->NewByteArray(jenv, sizeof(struct evmc_address));
+        assert(jaddress != NULL);
+        (*jenv)->SetByteArrayRegion(jenv, jaddress, 0, sizeof(struct evmc_address),
+                                    (jbyte*)address);
+
+        jdata = (*jenv)->NewByteArray(jenv, data_size);
+        assert(jdata != NULL);
+        (*jenv)->SetByteArrayRegion(jenv, jdata, 0, data_size, (jbyte*)data);
+
+        jclass byte_type = (*jenv)->FindClass(jenv, "[B");
+        jtopics = (*jenv)->NewObjectArray(jenv, (jsize)topics_count, byte_type, NULL);
+        assert(jtopics != NULL);
+        for (int i = 0; i < topics_count; i++)
+        {
+            jbyteArray jtopic = (*jenv)->NewByteArray(jenv, BYTES32_SIZE);
```

Comment:
Please use `sizeof(struct evmc_bytes32)`, makes it easier to grep.

Reply:
done